packages/storage/src/service.ts

```diff
@@ -0,0 +1,238 @@
+    const headObjectCommand = new HeadObjectCommand({
+      Bucket: S3_BUCKET_NAME,
+      Key: fileKey,
+    });
+
+    try {
+      await s3Client.send(headObjectCommand);
+    } catch (error: unknown) {
+      logger.error({ error }, "Failed to check if file exists");
+      if (
+        (error as Error).name === "NotFound" ||
+        (error as { $metadata?: { httpStatusCode?: number } }).$metadata?.httpStatusCode === 404
+      ) {
+        return err({
+          code: ErrorCode.FileNotFoundError,
+        });
+      }
+
+      logger.warn({ error, fileKey }, "HeadObject check failed; proceeding to sign download URL");
+    }
+
+    const getObjectCommand = new GetObjectCommand({
+      Bucket: S3_BUCKET_NAME,
+      Key: fileKey,
+    });
+
+    return ok(await getSignedUrl(s3Client, getObjectCommand, { expiresIn: 60 * 30 }));
+  } catch (error) {
+    logger.error({ error }, "Failed to get signed download URL");
+    return err({
+      code: ErrorCode.Unknown,
+    });
+  }
+};
+
+/**
+ * Delete a file from S3
+ * @param fileKey - The key of the file in S3 (e.g. "surveys/123/responses/456/file.pdf")
+ * @returns A Result containing the void or an error: StorageError
+ */
+export const deleteFile = async (fileKey: string): Promise<Result<void, StorageError>> => {
+  try {
+    if (!s3Client) {
+      return err({
+        code: ErrorCode.S3ClientError,
+      });
+    }
+
+    if (!S3_BUCKET_NAME) {
+      return err({
+        code: ErrorCode.S3CredentialsError,
+      });
+    }
+
+    const deleteObjectCommand = new DeleteObjectCommand({
+      Bucket: S3_BUCKET_NAME,
+      Key: fileKey,
+    });
+
+    await s3Client.send(deleteObjectCommand);
+
+    return ok(undefined);
+  } catch (error) {
+    logger.error({ error }, "Failed to delete file");
+
+    return err({
+      code: ErrorCode.Unknown,
+    });
+  }
+};
+
+export const deleteFilesByPrefix = async (prefix: string): Promise<Result<void, StorageError>> => {
+  try {
+    if (!s3Client) {
+      return err({
+        code: ErrorCode.S3ClientError,
+      });
+    }
+
+    if (!S3_BUCKET_NAME) {
+      return err({
+        code: ErrorCode.S3CredentialsError,
+      });
+    }
+
+    const keys: { Key: string }[] = [];
+    let continuationToken: string | undefined;
+
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition, no-constant-condition -- We want to loop until the continuation token is undefined
+    while (true) {
+      const listObjectsCommand = new ListObjectsV2Command({
+        Bucket: S3_BUCKET_NAME,
+        Prefix: prefix,
+        ContinuationToken: continuationToken,
+      });
+
+      const page = await s3Client.send(listObjectsCommand);
+      const pageKeys = page.Contents?.flatMap((obj) => (obj.Key ? [{ Key: obj.Key }] : [])) ?? [];
+      keys.push(...pageKeys);
+
+      continuationToken = page.IsTruncated ? page.NextContinuationToken : undefined;
+
+      if (!continuationToken) {
+        break;
+      }
+    }
```

Comment:
🔧 **Fix**: We can use the built-in paginator here:

```suggestion
// Use paginator to automatically handle pagination
    const paginator = paginateListObjectsV2(
      { client: s3Client },
      {
        Bucket: S3_BUCKET_NAME,
        Prefix: prefix,
      }
    );

    for await (const page of paginator) {
      const pageKeys = page.Contents?.flatMap((obj) => (obj.Key ? [{ Key: obj.Key }] : [])) ?? [];
      keys.push(...pageKeys);
    }
```

Reply:
wow had no idea this existed, thanks for the suggestion! 🙌